# Scrape Bordeaux Wine Reviews from winespectator.com
## 1. Perform Login (This is not a hacking)
### You'll need a username and password!!!
### 1.   Go to login page

  https://www.winespectator.com/auth/login?returnpage=https://www.winespectator.com/
  In Google Chrome, do Ctrl+Shit+I (Inspect)
  ![alt text](https://drive.google.com/uc?id=11tmBcmdTz9oAKOLQCesAkqUUNAX3eW6n)

### 2.   Enter your username and password
![alt text](https://drive.google.com/uc?id=1jBgdSydU4_yDM9xFwY0kccdkPd3QB4yx)


The information we need is the **cookie** and **User-Agent**





  


In [26]:
import requests
from bs4 import BeautifulSoup

import re


def login(url):
    #url = 'https://www.winespectator.com/wine/search?submitted=Y&scope=ratings&winery=bordeaux'
    #url='https://www.winespectator.com/wine/search?submitted=Y&page=1&winery=bordeaux&text_search_flag=wine_plus_vintage&search_by=all&scorelow=-1&scorehigh=-1&pricelow=-1&pricehigh=-1&case_prod=null_case_prod&taste_date=&issue_date=&issue_year=&varietal%5B%5D=null_varietal&regions%5B%5D=null_regions&vintage%5B%5D=&size=15&sort_by=score&sort_dir=desc'
    header = {
        'Cookie': '_ga=GA1.2.1635341982.1567698554; _gid=GA1.2.1323968393.1567698554; __gads=ID=75b881e44fbdae2e:T=1567698553:S=ALNI_MYcKnm18GrtWj1kRJ37fu7nMN9fdA; _ceir=1; wso_session=1-5876b7247aef3669e64192c32d4ab050; bolt_session_3fd645b695cc401a1e18e1f62fad493b=ceb75294412307e0b26fcea6228957c9; __utma=93936667.1635341982.1567698554.1567698816.1567698816.1; __utmc=93936667; __utmz=93936667.1567698816.1.1.utmcsr=winespectator.com|utmccn=(referral)|utmcmd=referral|utmcct=/wine/search; xyz_cr_781_et_100==NaN&cr=781&et=100&ap=; wso_li=1; _gat=1',
        'User-Agent': 'Mozilla/5.0'

        }
    infor = requests.get(url, headers=header)

    return infor.text

def main(url,n):
    unifo1 =[]
    html = login(url)
    print(html)

main("https://www.winespectator.com/wine/search?submitted=Y&page=1&winery=bordeaux+2001&text_search_flag=wine_plus_vintage&search_by=all&scorelow=-1&scorehigh=-1&pricelow=-1&pricehigh=-1&case_prod=null_case_prod&taste_date=&issue_date=&issue_year=&varietal%5B%5D=null_varietal&regions%5B%5D=null_regions&vintage%5B%5D=&size=15&sort_by=score&sort_dir=desc",15)
    


<!DOCTYPE html>
<!--[if lt IE 7]>      <html lang="en"  class="no-js lt-ie9 lt-ie8 lt-ie7"> <![endif]-->
<!--[if IE 7]>         <html lang="en"  class="no-js lt-ie9 lt-ie8"> <![endif]-->
<!--[if IE 8]>         <html lang="en"  class="no-js lt-ie9"> <![endif]-->
<!--[if gt IE 8]><!--> <html lang="en" class="no-js"> <!--<![endif]-->
    <head>

            <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

    
    <script>
  window.dataLayer = window.dataLayer || [];
  window.dataLayer.push({
   'channel' : 'search',
   'member_type' : 'SUB',
   'channel_top' : 'wineratings'
  });
</script>
    <title>Wine Ratings Search Results | Wine Spectator</title>
    <meta name="title" content="Wine Ratings Search Results | Wine Spectator" />
<meta name="robots" content="index, follow" />
<meta name="description" content="Wine Spectator" />
<meta name="language" content="en" /

#2. Retrive information from a single page

## 2.1  Information we need

1.   Wine Name
2.   Wine Review
3.   Wine Country -- All our wines are from Bordeaux, France. So we don't need this
4.   Year  Score  Price 

## 2.2  Look for the locations! 
![alt text](https://drive.google.com/uc?id=1pDDebfnwsNywiFeHIkkxtiVBdCiXA-3z)

2.2.1.   Wine Name

![](https://drive.google.com/uc?id=1BfpUQ2aYnd9TRlqvR4--N52f89a1d4R4)





In [29]:
def fillWinedata(Winedata,html):
    wine_infors =[]
    soup = BeautifulSoup(html, 'html.parser')

    #print(num_wine)
    #Get the name of the wine
    wine_names = soup.find_all('p', 'm-0')
    print(wine_names[0].get_text())
    print(wine_names[0].get_text(' ','br/'))
def main(url,n):
    unifo1 =[]
    html = login(url)
    #print(html)
    fillWinedata(unifo1,html)
    #print(a)
    #printList(unifo1,n)

main("https://www.winespectator.com/wine/search?submitted=Y&page=1&winery=bordeaux+2001&text_search_flag=wine_plus_vintage&search_by=all&scorelow=-1&scorehigh=-1&pricelow=-1&pricehigh=-1&case_prod=null_case_prod&taste_date=&issue_date=&issue_year=&varietal%5B%5D=null_varietal&regions%5B%5D=null_regions&vintage%5B%5D=&size=15&sort_by=score&sort_dir=desc",15)

Château Doisy DaëneBarsac L'Extravagant
Château Doisy Daëne Barsac L'Extravagant


# 2.2.2   Wine Review

![alt text](https://drive.google.com/uc?id=1YprBO8qu31m8nghIftzBmnqUcxwFdHHy)


# 2.2.3   Wine Country 

![alt text](https://drive.google.com/uc?id=1gMEMsmzlQklPwisiIdy7aKLamNvbLjKH)




In [48]:
def fillWinedata(Winedata,html):
    wine_infors =[]
    soup = BeautifulSoup(html, 'html.parser')
    #Get the reviews
    wine_reviews = soup.find_all('div','collapse collapse-tn pb-16')
    print(wine_reviews[0])
def main(url,n):
    unifo1 =[]
    html = login(url)
    #print(html)
    fillWinedata(unifo1,html)
    #print(a)
    #printList(unifo1,n)

main("https://www.winespectator.com/wine/search?submitted=Y&page=1&winery=bordeaux+2001&text_search_flag=wine_plus_vintage&search_by=all&scorelow=-1&scorehigh=-1&pricelow=-1&pricehigh=-1&case_prod=null_case_prod&taste_date=&issue_date=&issue_year=&varietal%5B%5D=null_varietal&regions%5B%5D=null_regions&vintage%5B%5D=&size=15&sort_by=score&sort_dir=desc",15)

<div class="collapse collapse-tn pb-16" id="v145393">Liquid honey in appearance. Incredibly ripe with dried apricot, orange and mace. Full-bodied, thick and powerful with amazing richness and spiciness. It goes on and on and on. This concentration is phenomenal. Yet it's lively and spicy. Huge finish. Best after 2012. 140 cases made.  <em>–JS</em><div class="paragraph"><strong>Country:</strong> France • <strong>Region:</strong> Bordeaux • <strong>Issue Date:</strong> Sep 15, 2004</div><div class="mb-0 mt-16 d-md-none form-group form-check custom-checkbox"><input id="wineid_in_145393" name="wineid_in[]" type="checkbox" value="145393"/> <label for="wineid_in_145393">Select to Add</label></div></div>


## What have returned above:
Liquid honey in appearance. Incredibly ripe with dried apricot, orange and mace. Full-bodied, thick and powerful with amazing richness and spiciness. It goes on and on and on. This concentration is phenomenal. Yet it's lively and spicy. Huge finish. Best after 2012. 140 cases made.  –JSCountry: France • Region: Bordeaux • Issue Date: Sep 15, 2004 Select to Add


**The Reviews is rolling up with the Country/ Region/ Issue Date???**

![alt text](https://drive.google.com/uc?id=1gZvLCQG5B0vCFKp1B3hJYZpD0C5SUliS)

In [55]:
def fillWinedata(Winedata,html):
    wine_infors =[]
    soup = BeautifulSoup(html, 'html.parser')

    #print(num_wine)
    #wine = soup.find_all('strong','text-uppercase')
    #Get the name of the wine
    wine_names = soup.find_all('p', 'm-0')
    #print(wine_names[0].get_text(' ','br/'))
    
    #Get the reviews
    wine_reviews = soup.find_all('div','collapse collapse-tn pb-16')
    print(wine_reviews[0].find(text=True))

    wine_judges = wine_reviews[0].find('em')
    print(wine_judges.get_text())
    #Get the Country/region/Issus Date
    wine_country = soup.find_all('div', 'paragraph')
    print(wine_country[0].get_text())

def main(url,n):
    unifo1 =[]
    html = login(url)
    fillWinedata(unifo1,html)
 

main("https://www.winespectator.com/wine/search?submitted=Y&page=1&winery=bordeaux+2001&text_search_flag=wine_plus_vintage&search_by=all&scorelow=-1&scorehigh=-1&pricelow=-1&pricehigh=-1&case_prod=null_case_prod&taste_date=&issue_date=&issue_year=&varietal%5B%5D=null_varietal&regions%5B%5D=null_regions&vintage%5B%5D=&size=15&sort_by=score&sort_dir=desc",15)


Liquid honey in appearance. Incredibly ripe with dried apricot, orange and mace. Full-bodied, thick and powerful with amazing richness and spiciness. It goes on and on and on. This concentration is phenomenal. Yet it's lively and spicy. Huge finish. Best after 2012. 140 cases made.  
–JS
Country: France • Region: Bordeaux • Issue Date: Sep 15, 2004



# 2.2.4   Year  Score  Price 

![alt text](https://drive.google.com/uc?id=1qK72D1mvSoDWZ0P20h-SLZEXPJ0Fjw5x)

In [41]:
def fillWinedata(Winedata,html):
    wine_infors =[]
    soup = BeautifulSoup(html, 'html.parser')
   # Get the Vintage/Score/Release
    wine_infor = soup.find_all('td', 'd-none d-md-table-cell border-0')
    print(wine_infor[0])
    print(wine_infor[1])
    print(wine_infor[2])
    print("The length of wine infor is : ",len(wine_infor))

    
def main(url,n):
    unifo1 =[]
    html = login(url)

    fillWinedata(unifo1,html)
 

main("https://www.winespectator.com/wine/search?submitted=Y&page=1&winery=bordeaux+2001&text_search_flag=wine_plus_vintage&search_by=all&scorelow=-1&scorehigh=-1&pricelow=-1&pricehigh=-1&case_prod=null_case_prod&taste_date=&issue_date=&issue_year=&varietal%5B%5D=null_varietal&regions%5B%5D=null_regions&vintage%5B%5D=&size=15&sort_by=score&sort_dir=desc",15)

<td class="d-none d-md-table-cell border-0">2001</td>
<td class="d-none d-md-table-cell border-0">100</td>
<td class="d-none d-md-table-cell border-0">$220/375ml</td>
The length of wine infor is :  45


## 15 wines! why the length of wines is 45???


In [45]:
def fillWinedata(Winedata,html):
    wine_infors =[]
    soup = BeautifulSoup(html, 'html.parser')
    wine_names = soup.find_all('p', 'm-0')
   # Get the Vintage/Score/Release
    wine_infor = soup.find_all('td', 'd-none d-md-table-cell border-0')
    print(wine_infor[0])
    print(wine_infor[1])
    print(wine_infor[2])
    print("The length of wine infor is : ",len(wine_infor))

     #seperate them
    a = 0
    for j in range(len(wine_names)):
        wine_infors.append([wine_infor[a].get_text(),wine_infor[a+1].get_text(),wine_infor[a+2].get_text()])
        a = a+3
    
    print(wine_infors[0][0])
    print(wine_infors[0][1])
    print(wine_infors[0][2])

def main(url,n):
    unifo1 =[]
    html = login(url)

    fillWinedata(unifo1,html)
 

main("https://www.winespectator.com/wine/search?submitted=Y&page=1&winery=bordeaux+2001&text_search_flag=wine_plus_vintage&search_by=all&scorelow=-1&scorehigh=-1&pricelow=-1&pricehigh=-1&case_prod=null_case_prod&taste_date=&issue_date=&issue_year=&varietal%5B%5D=null_varietal&regions%5B%5D=null_regions&vintage%5B%5D=&size=15&sort_by=score&sort_dir=desc",15)

<td class="d-none d-md-table-cell border-0">2001</td>
<td class="d-none d-md-table-cell border-0">100</td>
<td class="d-none d-md-table-cell border-0">$220/375ml</td>
The length of wine infor is :  45
2001
100
$220/375ml


# Oh! Yeah!!!

![alt text](https://drive.google.com/uc?id=1MuoOuZtW89HEpLNZMYOiskUBPfTEua-d)

Why score is BT???????

CHÂTEAU D'AIGUILHE QUERRE
Côtes de Castillon
A bit tough, but some good berry and wet earth character. Score range: 80-84   –JS
Country: France • Region: Bordeaux • Issue Date: Web Only - 2002

2001	BT	$NA

In [60]:
def fillWinedata(Winedata,html):
    wine_infors =[]
    soup = BeautifulSoup(html, 'html.parser')
    wine_names = soup.find_all('p', 'm-0')
    #Get the reviews
    wine_reviews = soup.find_all('div','collapse collapse-tn pb-16')
    # Get the Vintage/Score/Release
    wine_infor = soup.find_all('td', 'd-none d-md-table-cell border-0')
    print(wine_infor[0].get_text(), wine_infor[1].get_text(), wine_infor[2].get_text())
    

     #seperate them
    a = 0
    b = 0
    for j in range(len(wine_names)):

        if(wine_infor[a+1].get_text() == 'BT'):
            pattern = re.compile('Score range: (\d+)-(\d+)')
            num_wine = re.findall(pattern, wine_reviews[b].get_text())
            if(len(num_wine)!=0):
                avgerage_score = round(float((int(num_wine[0][0]) + int(num_wine[0][1])) / 2))
                wine_infors.append(
                    [wine_infor[a].get_text(), avgerage_score, wine_infor[a + 2].get_text()])
            else:

                wine_infors.append([wine_infor[a].get_text(), wine_infor[a + 1].get_text(), wine_infor[a + 2].get_text(',','-')])
        #     wine_infors.append([wine_infor[a].get_text(), round(float((int(num_wine[0][0])+int(num_wine[0][1]))/2),2), wine_infor[a + 2].get_text()])
        else:
            wine_infors.append([wine_infor[a].get_text(),wine_infor[a+1].get_text(),wine_infor[a+2].get_text()])
        a = a+3
        b = b+1
    print(wine_infors[0][0],wine_infors[0][1],wine_infors[0][2])

def main(url,n):
    unifo1 =[]
    html = login(url)

    fillWinedata(unifo1,html)
 

main("https://www.winespectator.com/wine/search/recall/yes/page/1/sort_by/score/sort_dir/asc",15)

2001 BT $NA
2001 82 $NA


## HAHAHAHAHA!

### We finallly scrape one page of data!


In [61]:
def fillWinedata(Winedata,html):
    wine_infors =[]
    soup = BeautifulSoup(html, 'html.parser')

    #print(num_wine)
    #wine = soup.find_all('strong','text-uppercase')
    #Get the name of the wine
    wine_names = soup.find_all('p', 'm-0')
    #print(wine_names[0].get_text(' ','br/'))
    
    #Get the reviews
    wine_reviews = soup.find_all('div','collapse collapse-tn pb-16')

    # Get the Vintage/Score/Release
    wine_infor = soup.find_all('td', 'd-none d-md-table-cell border-0')

     #Get the Country/region/Issus Date
    wine_country = soup.find_all('div', 'paragraph')
    #print(wine_country[0].get_text())

    #seperate them
    a = 0
    b = 0
    for j in range(len(wine_names)):

        if(wine_infor[a+1].get_text() == 'BT'):
            pattern = re.compile('Score range: (\d+)-(\d+)')
            num_wine = re.findall(pattern, wine_reviews[b].get_text())
            if(len(num_wine)!=0):
                avgerage_score = round(float((int(num_wine[0][0]) + int(num_wine[0][1])) / 2))
                wine_infors.append(
                    [wine_infor[a].get_text(), avgerage_score, wine_infor[a + 2].get_text()])
            else:

                wine_infors.append([wine_infor[a].get_text(), wine_infor[a + 1].get_text(), wine_infor[a + 2].get_text(',','-')])
        #     wine_infors.append([wine_infor[a].get_text(), round(float((int(num_wine[0][0])+int(num_wine[0][1]))/2),2), wine_infor[a + 2].get_text()])
        else:
            wine_infors.append([wine_infor[a].get_text(),wine_infor[a+1].get_text(),wine_infor[a+2].get_text()])
        a = a+3
        b = b+1

    for i in range(len(wine_names)):
        Winedata.append([wine_names[i].get_text(' ','br/'), wine_infors[i],wine_reviews[i].find(text=True).strip(),wine_reviews[i].find('em').get_text(),'Country: France'])


    return Winedata

def printList(Winedata,num):
    
    for i in range(num):
        u = Winedata[i]
        print("{}\n{}\t{}\n{}\n{}\t{}\t{}".format(u[0],u[2],u[3],u[4],u[1][0],u[1][1],u[1][2]))

def main(url,n):
    unifo1 =[]
    html = login(url)
    #print(html)
    fillWinedata(unifo1,html)
    #print(a)
    printList(unifo1,n)

main("https://www.winespectator.com/wine/search?submitted=Y&page=1&winery=bordeaux+2001&text_search_flag=wine_plus_vintage&search_by=all&scorelow=-1&scorehigh=-1&pricelow=-1&pricehigh=-1&case_prod=null_case_prod&taste_date=&issue_date=&issue_year=&varietal%5B%5D=null_varietal&regions%5B%5D=null_regions&vintage%5B%5D=&size=15&sort_by=score&sort_dir=desc",15)


Château Doisy Daëne Barsac L'Extravagant
Liquid honey in appearance. Incredibly ripe with dried apricot, orange and mace. Full-bodied, thick and powerful with amazing richness and spiciness. It goes on and on and on. This concentration is phenomenal. Yet it's lively and spicy. Huge finish. Best after 2012. 140 cases made.	–JS
Country: France
2001	100	$220/375ml
Château Rieussec Sauternes
Like lemon curd on the nose, turning to honey and caramel. Full-bodied and very sweet, with fantastic concentration of ripe and botrytized fruit, yet balanced and refined. Electric acidity. Lasts for minutes on the palate. This is absolutely mind-blowing. This is the greatest young Sauternes I have ever tasted. Best after 2010. 12,500 cases made.	–JS
Country: France
2001	100	$80
Château d'Yquem Sauternes
The greatest young Yquem I have ever tasted from bottle. Yellow, with a golden hue and an almost green tint. Intense aromas of botrytis, spices and blanched almonds follow through to honey, maple syrup

# 3  Retrive information from mutilple pages

## 3.1.   The secret is in the URL:
https://www.winespectator.com/wine/search?submitted=Y&page=1&winery=bordeaux+2001&text_search_flag=wine_plus_vintage&search_by=all&scorelow=-1&scorehigh=-1&pricelow=-1&pricehigh=-1&case_prod=null_case_prod&taste_date=&issue_date=&issue_year=&varietal%5B%5D=null_varietal&regions%5B%5D=null_regions&vintage%5B%5D=&size=15&sort_by=score&sort_dir=asc

## 3.2.   How many pages for each year???
![alt text](https://drive.google.com/uc?id=1oy--VWEKr9LiXAm62BmXzx-q6bsG2awN)





In [69]:
def get_numwine(html):
    soup = BeautifulSoup(html, 'html.parser')
    num_wines = soup.find_all('h3', 'float-md-left font-weight-normal mb-0')
    a = num_wines[0].get_text()
    a = a.replace(',','')
    pattern = re.compile('(\d+) wines found')
    num_wine = re.findall(pattern, a)

    return num_wine[0]
def getAllPages(url):
    html = login(url)
    a = get_numwine(html)
    return a
def getAllPages1():
    pages =[]
    for i in range(2000,2017):
        a = getAllPages("https://www.winespectator.com/wine/search?submitted=Y&page=1&winery=bordeaux+"+str(i)+"&text_search_flag=wine_plus_vintage&search_by=all&scorelow=-1&scorehigh=-1&pricelow=-1&pricehigh=-1&case_prod=null_case_prod&taste_date=&issue_date=&issue_year=&varietal%5B%5D=null_varietal&regions%5B%5D=null_regions&vintage%5B%5D=&size=15&sort_by=score&sort_dir=desc")
        pages.append(int(a))
    return pages

print(getAllPages1())



[710, 628, 537, 753, 666, 1073, 910, 491, 874, 1292, 1288, 1092, 998, 573, 853, 837, 774]


In [70]:
pages = [710, 628, 537, 753, 666, 1073, 910, 491, 874, 1292, 1288, 1092, 998, 573, 853, 837,774]
year = 2000
for j in pages:
    print("I am in year ", + year)
    for  i in range(int(j/15)):
        main("https://www.winespectator.com/wine/search?submitted=Y&page="+str(i+1)+"&winery=bordeaux+"+str(year)+"&text_search_flag=wine_plus_vintage&search_by=all&scorelow=-1&scorehigh=-1&pricelow=-1&pricehigh=-1&case_prod=null_case_prod&taste_date=&issue_date=&issue_year=&varietal%5B%5D=null_varietal&regions%5B%5D=null_regions&vintage%5B%5D=&size=15&sort_by=score&sort_dir=desc",15)
        print("I am in page " + str(i+1))
    main("https://www.winespectator.com/wine/search?submitted=Y&page="+str(int(j/15)+1)+"&winery=bordeaux+"+str(year)+"&text_search_flag=wine_plus_vintage&search_by=all&scorelow=-1&scorehigh=-1&pricelow=-1&pricehigh=-1&case_prod=null_case_prod&taste_date=&issue_date=&issue_year=&varietal%5B%5D=null_varietal&regions%5B%5D=null_regions&vintage%5B%5D=&size=15&sort_by=score&sort_dir=desc",int(j%15))
    year = year+1


I am in year  2000
Château Léoville Las Cases St.-Julien
Absolutely fantastic. This is one of the most exciting young reds I have tasted in a long, long time. It shows intense aromas of berries, currants and minerals, with hints of mint. Full-bodied and packed with fruit and tannins, its long finish is refined and silky. A benchmark for the vintage. Las Cases has always wanted to make first-growth quality in a top-notch vintage, and it certainly did in 2000. Best after 2012. 15,000 cases made.	–JS
Country: France
2000	100	$170
Château Latour Pauillac
The fruit here is still very much in the primary phase, with a decidedly racy feel to the raspberry coulis, cassis and blackberry reduction notes that are streaked with violet, iron and graphite flavors. The superlong finish alternates between a tug of sweet earth and a velvety feel, as the fruit and grip are still melding together, but there's so much vivacity here, there's no concern with waiting it out. The wait may be a while though. R

KeyboardInterrupt: ignored